nexus/src/app/sagas/instance_create.rs

```diff
@@ -470,6 +481,40 @@ async fn sic_create_network_interfaces_undo(
     Ok(())
 }
 
+/// Create an external IP address for the instance.
+async fn sic_allocate_external_ip(
+    sagactx: ActionContext<SagaInstanceCreate>,
+) -> Result<Uuid, ActionError> {
+    let osagactx = sagactx.user_data();
+    let datastore = osagactx.datastore();
+    let saga_params = sagactx.saga_params();
+    let opctx =
+        OpContext::for_saga_action(&sagactx, &saga_params.serialized_authn);
+    let instance_id = sagactx.lookup::<Uuid>("instance_id")?;
+    let external_ip = datastore
+        .allocate_instance_external_ip(&opctx, instance_id)
```

Comment:
Double-checking: Since this is an action in a saga, is this idempotent?

Same question for deallocate.

Reply:
Allocation is not idempotent, no. I think it's pretty straightforward to do, though, so I'll take a crack now.

I'm also going to make the method `deallocate_instance_external_ip` idempotent, since (1) that's easy and (2) it's the one used in sagas. The method `deallocate_instance_external_ip_by_instance_id` is more difficult to make idempotent. The `UpdateAndCheck` trait doesn't apply, since the query looks up data by a key other than the primary key, and so returns potentially many results. It's not used in sagas, however, so I'm less concerned about it now. I'll make an issue for at least that one, and maybe the allocation if I can't make quick work of it.

Thanks for flagging this.

Reply:
Ok, deallocation is now idempotent. I'm going to defer allocation too, in the interest of urgency. It's possible to make the action idempotent, by requiring the primary key be passed as an argument. We do this for network interfaces for example: generate UUIDs in one saga, use them in the next, and detect / ignore primary key uniqueness violations.

That would work here, but we can't get back the record itself (which the function returns) without a second query, opening a race window. That's one step forward, and one back. I'd rather just flag this as something to come back to, requiring some more extensive work, such as writing a query to return a new or existing record, or returning a different type that expresses this specific situation. Sagas would ignore that, other functions would raise hell. In any case, deferred for now, I'll make an issue.